clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -3565,6 +3565,82 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
     return RValue::get(emitBuiltinObjectSize(E->getArg(0), Type, ResType,
                                              /*EmittedE=*/nullptr, IsDynamic));
   }
+  case Builtin::BI__builtin_counted_by_ref: {
+    llvm::Value *Result = llvm::ConstantPointerNull::get(
+        cast<llvm::PointerType>(ConvertType(E->getType())));
+
+    const Expr *Arg = E->getArg(0)->IgnoreParenImpCasts();
+
+    if (const auto *UO = dyn_cast<UnaryOperator>(Arg);
+        UO && UO->getOpcode() == UO_AddrOf) {
+      Arg = UO->getSubExpr()->IgnoreParenImpCasts();
+
+      if (const auto *ASE = dyn_cast<ArraySubscriptExpr>(Arg))
+        Arg = ASE->getBase()->IgnoreParenImpCasts();
+    }
+
+    if (const auto *ME = dyn_cast_if_present<MemberExpr>(Arg)) {
+      ASTContext &Ctx = getContext();
+      bool IsFlexibleArrayMember = ME->isFlexibleArrayMemberLike(
+          Ctx, getLangOpts().getStrictFlexArraysLevel());
+
+      if (!ME->HasSideEffects(Ctx) && IsFlexibleArrayMember &&
+          ME->getMemberDecl()->getType()->isCountAttributedType()) {
+        FieldDecl *FAMDecl = cast<FieldDecl>(ME->getMemberDecl());
+        if (FieldDecl *CountFD = FAMDecl->findCountedByField()) {
+          QualType CountTy = CountFD->getType();
+          SmallVector<NamedDecl *, 2> PathToFD;
+
+          // Reverse through any anonymous structs / unions surrounding the
+          // flexible array member. We'll build any necessary MemberExpr's to
+          // anonymous structs / unions when building a reference to the
+          // 'count' field.
+          RecordDecl *RD = FAMDecl->getParent();
+          DeclContext *DC = RD;
+          for (; DC->isRecord(); DC = DC->getLexicalParent()) {
+            if (!RD->isAnonymousStructOrUnion())
+              break;
+            RD = cast<RecordDecl>(DC);
+            if (auto *Base = dyn_cast<MemberExpr>(ME->getBase()))
+              ME = Base;
+          }
+
+          // See if the count's FieldDecl is within anonymous structs.
+          for (Decl *D : RD->decls()) {
+            if (auto *IFD = dyn_cast<IndirectFieldDecl>(D);
+                IFD && IFD->getAnonField() == CountFD) {
+              PathToFD.insert(PathToFD.begin(), IFD->chain_begin(),
+                              IFD->chain_end());
+              break;
+            }
+          }
+
+          if (PathToFD.empty())
+            PathToFD.push_back(CountFD);
+
+          // Build a MemberExpr to the 'count' field. This accounts for any
+          // anonymous structs / unions that may contain the field.
+          bool isArrow = ME->isArrow();
+          Expr *New = ME->getBase();
+          for (NamedDecl *ND : PathToFD) {
+            ValueDecl *VD = cast<ValueDecl>(ND);
+            New = MemberExpr::CreateImplicit(
```

Comment:
I'm not sure we are encouraged to create AST nodes in CodeGen. Was there a reason that it's difficult to do as a direct IRGen?

Reply:
If you're adamant that I can't generate AST nodes here, I could do it in Sema like you suggested in the RFC.

Reply:
No, I don't have a strong opinion about it myself. I guess the Clang code owners might want to look at it though @AaronBallman @rjmccall .

Reply:
I don't think we should be creating AST nodes during CodeGen -- that will be very confusing for things which consume the AST and expect it to be semantically accurate (such as clang-tidy checks or anything using AST matchers, really).

CC @efriedma-quic as well

Reply:
Who consumes the AST after code gen?

I assume you'd like me to generate the AST nodes during sema then?

Reply:
> Who consumes the AST after code gen?

Nobody, which is the point. We want the AST to be complete so that handing it off to something else to process works the same way whether you're doing CodeGen or walking the AST for other purposes.

> I assume you'd like me to generate the AST nodes during sema then?

Correct.

> I could also delete the nodes I created after code generate as they're no longer needed. Would that help?

Not really; the critical part is representing the correct semantics in the AST itself. Generating the AST nodes during CodeGen and then acting on those nodes is where the problem comes from -- that means the AST is missing information that's important for the semantics.

Reply:
Okay. Done.